encoding/kzg/prover/icicle/multiframe_proof.go

```diff
@@ -0,0 +1,212 @@
+	"github.com/Layr-Labs/eigenda/encoding/icicle"
+	"github.com/Layr-Labs/eigenda/encoding/kzg"
+	"github.com/Layr-Labs/eigenda/encoding/utils/toeplitz"
+	"github.com/consensys/gnark-crypto/ecc/bn254"
+	"github.com/consensys/gnark-crypto/ecc/bn254/fr"
+	"github.com/ingonyama-zk/icicle/v3/wrappers/golang/core"
+	iciclebn254 "github.com/ingonyama-zk/icicle/v3/wrappers/golang/curves/bn254"
+	"github.com/ingonyama-zk/icicle/v3/wrappers/golang/runtime"
+)
+
+type KzgMultiProofIcicleBackend struct {
+	*kzg.KzgConfig
+	Fs             *fft.FFTSettings
+	FlatFFTPointsT []iciclebn254.Affine
+	SRSIcicle      []iciclebn254.Affine
+	SFs            *fft.FFTSettings
+	Srs            *kzg.SRS
+	NttCfg         core.NTTConfig[[iciclebn254.SCALAR_LIMBS]uint32]
+	MsmCfg         core.MSMConfig
+	Device         runtime.Device
+}
+
+type WorkerResult struct {
+	err error
+}
+
+// This function supports batching over multiple blobs.
+// All blobs must have same size and concatenated passed as polyFr
+func (p *KzgMultiProofIcicleBackend) ComputeMultiFrameProof(polyFr []fr.Element, numChunks, chunkLen, numWorker uint64) ([]bn254.G1Affine, error) {
+	begin := time.Now()
+
+	// Robert: Standardizing this to use the same math used in precomputeSRS
+	dimE := numChunks
+	l := chunkLen
+	numPoly := uint64(len(polyFr)) / dimE / chunkLen
+
+	jobChan := make(chan uint64, numWorker)
+	results := make(chan WorkerResult, numWorker)
+
+	// create storage for intermediate fft outputs
+	coeffStore := make([][]fr.Element, dimE*2)
+	for i := range coeffStore {
+		coeffStore[i] = make([]fr.Element, l)
+	}
+
+	for w := uint64(0); w < numWorker; w++ {
+		go p.proofWorker(polyFr, jobChan, l, dimE, coeffStore, results)
+	}
+
+	for j := uint64(0); j < l; j++ {
+		jobChan <- j
+	}
+	close(jobChan)
+
+	// return last error
+	var err error
+	for w := uint64(0); w < numWorker; w++ {
+		wr := <-results
+		if wr.err != nil {
+			err = wr.err
+		}
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("proof worker error: %v", err)
+	}
+
+	preprocessDone := time.Now()
+
+	flattenCoeffStoreFr := make([]fr.Element, 0)
+	for i := 0; i < len(coeffStore); i++ {
+		flattenCoeffStoreFr = append(flattenCoeffStoreFr, coeffStore[i]...)
+	}
+
+	flattenCoeffStoreSf := icicle.ConvertFrToScalarFieldsBytes(flattenCoeffStoreFr)
+	flattenCoeffStoreCopy := core.HostSliceFromElements[iciclebn254.ScalarField](flattenCoeffStoreSf)
+
+	var icicleFFTBatch []bn254.G1Affine
+	var icicleErr error
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	var msmDone, firstECNttDone, secondECNttDone time.Time
+	runtime.RunOnDevice(&p.Device, func(args ...any) {
+		defer wg.Done()
+
+		// Copy the flatten coeff to device
+		var flattenStoreCopyToDevice core.DeviceSlice
+		flattenCoeffStoreCopy.CopyToDevice(&flattenStoreCopyToDevice, true)
+
+		sumVec, err := p.MsmBatchOnDevice(flattenStoreCopyToDevice, p.FlatFFTPointsT, int(numPoly)*int(dimE)*2)
+		if err != nil {
+			icicleErr = fmt.Errorf("msm error: %w", err)
```

Comment:
here and below: why not return on such an error? The proof computing is already failed at these points